### PR TITLE
Stabilize cfg(doc)

### LINKED ISSUE #43781: Tracking issue for `#[doc(cfg(...))]`
This is a tracking issue for the `#[doc(cfg(...))]` attribute introduced in #43348.

**Steps:**

- [x] Implement
- [ ] Adjust documentation ([see instructions on forge][doc-guide])
- [ ] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

(cc #1998)

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
cc #43781.

### CODE REVIEW

File: src/doc/rustdoc/src/the-doc-attribute.md
Comment (Centril) [inline src/doc/rustdoc/src/the-doc-attribute.md]: How does this cope with functions that have `-> impl Trait` where `!` does not implement `Trait`?

File: src/doc/rustdoc/src/the-doc-attribute.md
Comment (Manishearth) [inline src/doc/rustdoc/src/the-doc-attribute.md]: Unsure if this implementation detail needs to be exposed.

**Review by Manishearth — APPROVED** [review]

File: src/doc/rustdoc/src/the-doc-attribute.md
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/the-doc-attribute.md]: cc @QuietMisdreavus

File: src/doc/rustdoc/src/the-doc-attribute.md
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/the-doc-attribute.md]: I found it interesting so I didn't remove it. :)

File: src/doc/rustdoc/src/the-doc-attribute.md
Comment (ollie27) [reply src/doc/rustdoc/src/the-doc-attribute.md]: There is currently an exception for `-> impl Trait`: https://github.com/rust-lang/rust/blob/26451ef7b5e00887dc8f27717ff34262df23d655/src/librustc_interface/util.rs#L668-L673

This means that `#[doc(cfg(...))]` doesn't work for those functions yet so I don't think this feature is ready for stabilization yet.

**Comment by GuillaumeGomez** [comment]
Forgot to remove a feature test. Updated.

**Comment by Centril** [comment]
@bors r- 

There's an ongoing discussion in https://github.com/rust-lang/rust/pull/61351#discussion_r289109410 and a T-rustdoc team member who things this is not ready to stabilize yet.

File: src/doc/rustdoc/src/the-doc-attribute.md
Comment (Centril) [reply src/doc/rustdoc/src/the-doc-attribute.md]:
```rust
//  * `impl Trait`, due to issue #43869 that functions returning impl Trait cannot be diverging.
//    Solving this may require `!` to implement every trait, which relies on the an even more
//    ambitious form of the closed RFC #1637. See also [#34511].
```
This bit seems unlikely from my POV as a T-Lang member... cc @rust-lang/lang if y'all have thoughts...

Comment (QuietMisdreavus) [reply src/doc/rustdoc/src/the-doc-attribute.md]: I'm with @ollie27 here, i had assumed the reason we couldn't stabilize `doc(cfg)` yet was because the implementation didn't work for all situations.

**Comment by QuietMisdreavus** [comment]
I'm against stabilizing `doc(cfg)`, because of the following snippet:

```rust
#[cfg(any(rustdoc, windows))]
#[doc(cfg(windows))]
pub fn my_handle() -> winapi::shared::ntdef::HANDLE { ... }
```

We currently don't have a way to handle this code in rustdoc if it's being compiled on non-windows targets, because the `winapi` crate is totally empty on non-windows targets. This sample will create a type error when rustdoc tries to build it. This sample was discussed at the Rust All-Hands this year, but as far as i know we haven't acted on the discussion that occurred there. In addition, as @ollie27 mentioned, functions returning `impl Trait` do not receive the benefits of `everybody_loops`, so the compilation is not perfect in those cases.

-----

...however, i *would* like to stabilize `cfg(rustdoc)` separately, since that doesn't have the same ramifications of the rest of `doc(cfg)`. My recommendation would be to either strip out `doc(cfg)` from this PR and just keep `cfg(rustdoc)`, or close this and open a new one that just stabilizes `cfg(rustdoc)`.

**Comment by GuillaumeGomez** [comment]
I initially wanted to have only `cfg(rustdoc)` but saw that the two were "linked" so I just stabilized both. I clearly don't mind to only stabilize `cfg(rustdoc)` (quite the opposite!). I'll update the PR soon.

**Comment by GuillaumeGomez** [comment]
It now only stabilizes `cfg(rustdoc)`.

**Comment by GuillaumeGomez** [comment]
ping @QuietMisdreavus @ollie27

File: src/doc/rustdoc/src/the-doc-attribute.md
Comment (QuietMisdreavus) [inline src/doc/rustdoc/src/the-doc-attribute.md]: These samples should just use `#[cfg(any(platform, rustdoc))]`, to match the text above it. As written, they will still only appear when generating the docs using a `--target` for their selected platform. (The current cfg attributes are like, "this should only appear *either* on Windows/Unix, *or* when *both* Windows/Unix and `rustdoc` are set".)

File: src/test/ui/cfg-rustdoc.rs
Comment (QuietMisdreavus) [inline src/test/ui/cfg-rustdoc.rs]: I wonder if this sample should use `#[deny(dead_code)]` (or whatever warning trips on unused structs) to ensure that regular compilation doesn't even notice the struct.

Is there an existing test that ensures that item with `cfg(rustdoc)` appear in documentation? I couldn't find any on a quick scan - it doesn't seem like the `doc(cfg)` tests use it. We should make sure that we have a test for it.

**Review by QuietMisdreavus — COMMENTED** [review]
Thanks for reworking the PR! I have a couple comments.

I see the documentation you added is part of what was written for `doc(cfg)`. We should probably rework the documentation there to link to this section as well. We should also take out the mention of `cfg(rustdoc)` in the Unstable Book's page for `cfg_doc` as well, since that feature flag no longer governs this.

**Comment by GuillaumeGomez** [comment]
> I see the documentation you added is part of what was written for `doc(cfg)`. We should probably rework the documentation there to link to this section as well. We should also take out the mention of `cfg(rustdoc)` in the Unstable Book's page for `cfg_doc` as well, since that feature flag no longer governs this.

It's used as example and I find useful. Are you sure you want me to remove it? Having it there seems pretty nice to me actually.

File: src/doc/rustdoc/src/the-doc-attribute.md
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/the-doc-attribute.md]: Agreed, that even simplifies the example. Updating it.

**Comment by QuietMisdreavus** [comment]
> > I see the documentation you added is part of what was written for `doc(cfg)`. We should probably rework the documentation there to link to this section as well. We should also take out the mention of `cfg(rustdoc)` in the Unstable Book's page for `cfg_doc` as well, since that feature flag no longer governs this.
> 
> It's used as example and I find useful. Are you sure you want me to remove it? Having it there seems pretty nice to me actually.

You don't need to remove it entirely, just rephrase the text around it to make it stop saying that `#![feature(doc_cfg)]` enables `cfg(rustdoc)`, or to briefly mention `cfg(rustdoc)`, link to this new section, and leave the example there.

File: src/test/ui/cfg-rustdoc.rs
Comment (GuillaumeGomez) [reply src/test/ui/cfg-rustdoc.rs]: We have one in `src/test/ui/cfg-rustdoc.rs`. :)

**Comment by GuillaumeGomez** [comment]
Ok, I think all have been done. :)

File: src/doc/rustdoc/src/unstable-features.md
Comment (QuietMisdreavus) [inline src/doc/rustdoc/src/unstable-features.md]: This was probably the wrong line to remove. Looking at these docs again, they can probably stay as-is, if we rephrase the `cfg(rustdoc)` ones and link to them for more information.

File: src/doc/rustdoc/src/the-doc-attribute.md
Comment (QuietMisdreavus) [inline src/doc/rustdoc/src/the-doc-attribute.md]: How about this for rewriting these docs:

> For conditional compilation, Rustdoc treats your crate the same way the compiler does: Only things from the host target are available (or from the given `--target` if present), and everything else is "filtered out" from the crate. This can cause problems if your crate is providing different things on different targets, and you want your documentation to reflect all the available items you provide.
>
> If you want to make sure an item is seen by Rustdoc regardless of what platform it's being run from, you can apply `#[cfg(rustdoc)]` to it. Rustdoc sets this whenever it's building documentation, so anything that uses that flag will make it into documentation it generates. To apply this to an item with other `#[cfg]` filters on it, you can write something like `#[cfg(any(windows, rustdoc))]`. This will preserve the item either when built normally on Windows, or when being documented anywhere. For example:
>
> (current code sample)
>
> Here, the respective tokens can only be used by dependent crates on their respective platforms, but they will both appear in documentation.

File: src/test/ui/cfg-rustdoc.rs
Comment (QuietMisdreavus) [reply src/test/ui/cfg-rustdoc.rs]: That's this file that you're adding in this PR. It doesn't test that an item appears in documentation, only that it doesn't break compilation. I'm not sure this test is actually doing anything useful, even - in fact, removing the `#[cfg(rustdoc)]` from the struct doesn't change anything about the output. Can you change this test so that `main` tries to use `Foo` and capture the compilation error? And please add a `rustdoc` test that uses `cfg(any(windows, rustdoc))` and ensures the item is present when running on Unix.

**Review by QuietMisdreavus — COMMENTED** [review]
Please update the Unstable Book (`src/doc/unstable-book/src/language-features/doc-cfg.md`) to remove the paragraph about `cfg(rustdoc)`.

File: src/doc/rustdoc/src/the-doc-attribute.md
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/the-doc-attribute.md]: I really prefer your version so I'll just use it as is if you don't mind.

**Comment by GuillaumeGomez** [comment]
Updated. I'm not the sure the change in the unstable docs is the one you had in mind @QuietMisdreavus ?

**Comment by QuietMisdreavus** [comment]
That edit to the Rustdoc Book's unstable chapter is good. [I still want you to modify the Unstable Book, though](https://github.com/rust-lang/rust/pull/61351#pullrequestreview-252541214):

> Please update the Unstable Book (`src/doc/unstable-book/src/language-features/doc-cfg.md`) to remove the paragraph about `cfg(rustdoc)`.

**Comment by GuillaumeGomez** [comment]
Done!

File: src/doc/rustdoc/src/the-doc-attribute.md
Comment (QuietMisdreavus) [inline src/doc/rustdoc/src/the-doc-attribute.md]: I just realized something about these docs. First, the heading here is a little misleading, since `cfg(rustdoc)` isn't just about platform-specific stuff. It's feasible that people could "invent" items that only exist for documentation, for whatever reason. However, writing something like "exposing items to documentation" feels *more* misleading, since it implies that you need that to add items to docs in the first place. I think this header text is acceptable, since platform-specific docs was the driving force for it, but it's something to consider.

Second, there isn't really a good chapter for these docs. This chapter is specifically about the `#[doc]` attribute, but this section is about something you add to the `#[cfg]` attribute. However, all the other chapters make even less sense to add this to. I'm wondering whether we should create a new chapter that talks about "advanced" features that are stable. I'm not sure what else would go with it, though... :thinking: Maybe we can just leave this here. I'm curious what @ollie27 and @steveklabnik think, though.

**Comment by QuietMisdreavus** [comment]
As far as i'm concerned, this PR is ready to go, barring the one comment i just left, so i'm going to open up the FCP.

@rust-lang/rustdoc This PR is pulling out the `cfg(rustdoc)` feature and stabilizing it by itself. The feature allows users to expose items directly to rustdoc, even if they might not otherwise be available. It used to be tied up in the `#[doc(cfg)]` feature, but we've determined that it has enough utility on its own to stabilize it separately.

@rfcbot fcp merge

@rfcbot concern where-do-the-docs-go

**Comment by chocol4te** [comment]
@ollie27 @onur Ping from triage regarding your reviews! :)

(I saw the RFC bot's comment, I don't think 2/4 counts as a majority?)

**Comment by GuillaumeGomez** [comment]
And I need to update this PR once again...

**Comment by onur** [comment]
:+1:

**Comment by GuillaumeGomez** [comment]
With @onur approval (thanks a lot!), it can go forward now. I'll update the PR then it's ready to go.

**Comment by ollie27** [comment]
Is there a reason this isn't `cfg(doc)`? Given that we have `#[test]` and `cfg(test)` I would expect `#[doc]` and `cfg(doc)`. I guess it doesn't matter too much though.

I've noticed that `cfg(rustdoc)` is set inside doctests. This doctest currently passes:

```rust
//! ```
//! #![feature(doc_cfg)]
//! assert!(cfg!(rustdoc));
//! ```
```

Is that intentional? It looks like a bug to me.

**Comment by GuillaumeGomez** [comment]
Originally, I made it `cfg(doc)` but I completely forgot the reason why we decided to change.

> I've noticed that `cfg(rustdoc)` is set inside doctests. This doctest currently passes:

To me it doesn't look like a bug?

**Comment by GuillaumeGomez** [comment]
Updated.

@ollie27: Anything blocking on your side or can it make this move forward?

**Comment by Mark-Simulacrum** [comment]
It seems very strange for the cfg flag to be set inside doc tests, and since I can't imagine a use case where that would be helpful, I think we shouldn't set it to minimize exposed surface area. All doc tests are guaranteed to be run only by rustdoc :)

**Comment by GuillaumeGomez** [comment]
I'll remove it then.

**Comment by GuillaumeGomez** [comment]
Done.

**Comment by abonander** [comment]
@GuillaumeGomez what's the status of this? I've run into a use-case for exposing a `test_util` mod for doctests. (In fact, can anyone tell me why rustdoc apparently doesn't set `#[cfg(test)]` for doctests?)

**Comment by GuillaumeGomez** [comment]
Still waiting for approval. And now I need to rebase once again.

> (In fact, can anyone tell me why rustdoc apparently doesn't set #[cfg(test)] for doctests?)

To put it more simply: https://github.com/rust-lang/rust/pull/63803

**Comment by abonander** [comment]
Ah, TL;DR: 
> We had previously set `cfg(test)` when collecting doctests, starting in #59940, but this caused issues with libcore that caused all its doctests to be ignored, so that was reverted in #61199

**Comment by GuillaumeGomez** [comment]
Finally updated. \o/

**Comment by JohnCSimon** [comment]
Ping from triage
@Manishearth  - From the conversation it looks like this is ready for review?
cc: @GuillaumeGomez 
Thanks.

**Comment by Manishearth** [comment]
I'm not comfortable reviewing this, I don't have enough context on the stabilization here.

@quietmisdreavus ?

**Comment by JohnCSimon** [comment]
Ping from triage
@ollie27 @Centril ?

File: src/librustdoc/test.rs
Comment (Mark-Simulacrum) [inline src/librustdoc/test.rs]: Hm, what is the intent of this change? Why can't `rustdoc` be passed to libtest here?

Comment (GuillaumeGomez) [reply src/librustdoc/test.rs]: It was a request to specifically not pass the `rustdoc` features when running tests from a previous review. Or did I miss something?

Comment (Mark-Simulacrum) [reply src/librustdoc/test.rs]: Uh, sure, but this seems to have nothing to do with `--cfg` unless I'm missing something? I would expect the test_args vector to be the arguments like `--nocapture` and such?

Regardless, I would want this to be done not here but where we create test_args.

Comment (GuillaumeGomez) [reply src/librustdoc/test.rs]: Fine by me!

**Comment by GuillaumeGomez** [comment]
Updated!

File: src/doc/rustdoc/src/the-doc-attribute.md
Comment (Mark-Simulacrum) [inline src/doc/rustdoc/src/the-doc-attribute.md]:
```suggestion
If you want to make sure an item is seen by Rustdoc regardless of what platform it's targeting,
```

**Review by Mark-Simulacrum — COMMENTED** [review]
I think this is good to go other than this last nit; we need @QuietMisdreavus to resolve the where-do-the-docs-go concern -- IMO, it's fine as-is, it's not critical we get the docs 100% perfect in this PR (or ever - living docs!)

**Comment by GuillaumeGomez** [comment]
Then once CI passed, I'll r+ it. :) Thanks for your review @Mark-Simulacrum !

**Comment by ollie27** [comment]
I don't like that this uses the rustdoc name because I don't see why this feature would be specific to rustdoc. I would expect any Rust doc generation tool to also set this config. I think `cfg(doc)` would be better, does anyone know of a reason it can't be that?

As far as the where-do-the-docs-go is concerned I think moving the docs for this feature to a new page would be the right thing to do.

Where did the change to not set `cfg(rustdoc)` inside doctests go?

**Comment by GuillaumeGomez** [comment]
> I don't like that this uses the rustdoc name because I don't see why this feature would be specific to rustdoc. I would expect any Rust doc generation tool to also set this config. I think `cfg(doc)` would be better, does anyone know of a reason it can't be that?

It's been debatted a while ago. To put it simply, it's less accurate/explicit than `cfg(rustdoc)`.

> Where did the change to not set `cfg(rustdoc)` inside doctests go?

@Mark-Simulacrum preferred not making this change here, so I removed it.

**Comment by Centril** [comment]
> It's been debatted a while ago. To put it simply, it's less accurate/explicit than `cfg(rustdoc)`.

I'd argue that `cfg(doc)` is accurate *enough* to understand the intent of its usage by people who are reading the code. I think @ollie27 makes a good point re. other tools.

**Comment by Mark-Simulacrum** [comment]
> > Where did the change to not set cfg(rustdoc) inside doctests go?
> @Mark-Simulacrum preferred not making this change here, so I removed it.

I intended my comments from what I remember as "this does not seem like the right way to do it" not "they should not be done in this PR." I am happy to review them as part of a separate PR, though.

**Comment by GuillaumeGomez** [comment]
I think it'd be better in its own PR. Once this PR is merged, I'll create an issue about it.

> I'd argue that `cfg(doc)` is accurate *enough* to understand the intent of its usage by people who are reading the code. I think @ollie27 makes a good point re. other tools.

I can't find where the discussion about this happened (a long time ago, not even sure on which platform it happened...). If anyone has an idea? I recall it being quite the debate at the time and we finally picked `cfg(rustdoc)`. But if we could take another look at the discussion, maybe that'd make some points more clear?

**Comment by Mark-Simulacrum** [comment]
I would be against stabilizing a feature only to then change it in a backwards-incompatible way, so we should definitely not be passing the equivalent of `--cfg rustdoc` to doctests prior to landing this PR.

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum Ok, then should I add back the specific changes about this?

**Comment by Mark-Simulacrum** [comment]
Sure -- last time I was concerned with the specific way that was being done. It also seems very reasonable to extract that feature work into a separate PR that we can land before this one does.

**Comment by QuietMisdreavus** [comment]
So i'm trying to catch up on this thread. It looks like we want to make sure of a couple more things before we finalize this...

I agree that we shouldn't be setting `cfg(rustdoc)` inside doctests, so i went ahead and approved https://github.com/rust-lang/rust/pull/64741. I want that to be merged ahead of FCP, though:

@rfcbot concern cfg-in-doctests

Next, we should decide whether we should rename the flag to `cfg(doc)`, and how:

@rfcbot concern cfg-doc-or-cfg-rustdoc

I'm not opposed to renaming the flag to `cfg(doc)`. Is it feasible to make that change as part of the stabilization (i.e. in this PR) or should we land it separately, before the stabilization lands?

For the existing concern `where-do-the-docs-go`, i agree with @ollie27 here:

> As far as where-do-the-docs-go is concerned I think moving the docs for this feature to a new page would be the right thing to do.

We should make a new "Advanced features" page which contains the existing docs being added by this PR, as well as an additional header paragraph mentioning that the features on that page fall outside the other headings. If you want, i can write it for you.

**Comment by GuillaumeGomez** [comment]
> Next, we should decide whether we should rename the flag to `cfg(doc)`

Considering we now have `cfg(doctest)`, I think it'd make more sense to have `cfg(doc)`.

**Comment by GuillaumeGomez** [comment]
For more coherence with `cfg(doctest)`, I renamed the cfg to `cfg(doc)`.

cc @rust-lang/rustdoc

**Comment by QuietMisdreavus** [comment]
Looks like https://github.com/rust-lang/rust/pull/64741 is merged, so that resolves that concern...

@rfcbot resolve cfg-in-doctests

Regarding the change to `cfg(doc)`, would it be better to change that ahead of time (and include some kind of transition warning) or is there enough precedent in changing a thing at the same time as its stabilization? I'm not comfortable doing that kind of change without making sure that we've done something like that before.

And the docs [still need to be moved](https://github.com/rust-lang/rust/pull/61351#issuecomment-538021901):

> For the existing concern `where-do-the-docs-go`, i agree with @ollie27 here:
> 
> > As far as where-do-the-docs-go is concerned I think moving the docs for this feature to a new page would be the right thing to do.
> 
> We should make a new "Advanced features" page which contains the existing docs being added by this PR, as well as an additional header paragraph mentioning that the features on that page fall outside the other headings. If you want, i can write it for you.

**Comment by QuietMisdreavus** [comment]
After talking with @GuillaumeGomez on Discord, i've gone ahead and pushed up a commit that moves the docs onto a new page. Therefore...

@rfcbot resolve where-do-the-docs-go

**Comment by QuietMisdreavus** [comment]
@GuillaumeGomez and i talked some more about the naming concern, and decided to rename the attribute outside this PR, just so we can dodge the "name change and stabilize at the same time" issue by renaming it ahead of time. He volunteered to open the PR.

**Comment by QuietMisdreavus** [comment]
Now that https://github.com/rust-lang/rust/pull/66166 is merged, that should give people time to adjust to the new cfg name ahead of its stabilization. Therefore...

@rfcbot resolve cfg-doc-or-cfg-rustdoc

Let's get the FCP started!

**Pull request merged**
